File: src/main/java/com/kmarzecki/communicator/security/JwtTokenService.java
package com.kmarzecki.communicator.security;

import com.kmarzecki.communicator.model.auth.Role;
import com.kmarzecki.communicator.util.DateTimeProvider;
import io.jsonwebtoken.*;
import lombok.AllArgsConstructor;
import org.springframework.messaging.simp.stomp.StompHeaderAccessor;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import javax.servlet.http.HttpServletRequest;
import java.util.Base64;
import java.util.Date;
import java.util.Objects;
import java.util.Set;

@Component
@AllArgsConstructor
public class JwtTokenService {
    private static final String ROLES = "BASIC_ROLE";
    private static final String SECRET_KEY = "NON_SAFE_SECRET";
    private static final String ENCODED_SECRET_KEY = Base64.getEncoder().encodeToString(SECRET_KEY.getBytes());
    private static final String TOKEN_HEADER_NAME = "token";
    private static final int VALID_TIME_MS = 3600000;
    private final UserDetailsServiceImpl userDetailsService;
    private final DateTimeProvider dateProvider;

    public String createToken(String username, Set<Role> set) {
        Claims claims = Jwts.claims().setSubject(username);
        claims.put(ROLES, set);
        Date now = dateProvider.currentDate();
        Date validity = new Date(now.getTime() + VALID_TIME_MS);
        return Jwts.builder()
                .setClaims(claims)
                .setIssuedAt(now)
                .setExpiration(validity)
                .signWith(SignatureAlgorithm.HS256, ENCODED_SECRET_KEY)
                .compact();
    }

    public Authentication getAuthentication(String token) {
        UserDetails userDetails = userDetailsService.loadUserByUsername(getUsername(token));
        return new UsernamePasswordAuthenticationToken(userDetails, "", userDetails.getAuthorities());
    }

    public String getUsername(String token) {
        return Jwts.parser().setSigningKey(ENCODED_SECRET_KEY).parseClaimsJws(token).getBody().getSubject();
    }

    public String extractToken(HttpServletRequest req) {
        return req.getHeader(TOKEN_HEADER_NAME);
    }

    public String extractToken(StompHeaderAccessor accessor) {
        return Objects.requireNonNull(accessor.getNativeHeader(TOKEN_HEADER_NAME)).get(0);
    }

    public boolean validateToken(String token) {
        try {
            Jws<Claims> claims = Jwts.parser().setSigningKey(ENCODED_SECRET_KEY).parseClaimsJws(token);
            return !claims.getBody().getExpiration().before(new Date());
        } catch (JwtException | IllegalArgumentException e) {
            throw new JwtException("Expired or invalid JWT token");
        }
    }
}

File: src/main/kotlin/com/kmarzecki/trollo/security/JwtTokenService.kt
package com.kmarzecki.trollo.security

import com.kmarzecki.trollo.model.auth.Role
import com.kmarzecki.trollo.util.DateTimeProvider
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import lombok.AllArgsConstructor
import org.springframework.messaging.simp.stomp.StompHeaderAccessor
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Component
import java.util.*
import javax.servlet.http.HttpServletRequest

@Component
@AllArgsConstructor
class JwtTokenService(
        val userDetailsService: UserDetailsServiceImpl,
        val dateProvider: DateTimeProvider
) {
    fun createToken(username: String?, set: Set<Role?>): String {
        val claims = Jwts.claims().setSubject(username)
        claims[ROLES] = set
        val now = dateProvider.currentDate()
        val validity = Date(now.time + VALID_TIME_MS)
        return Jwts.builder()
                .setClaims(claims)
                .setIssuedAt(now)
                .setExpiration(validity)
                .signWith(SignatureAlgorithm.HS256, ENCODED_SECRET_KEY)
                .compact()
    }

    fun getAuthentication(token: String?): Authentication {
        val userDetails = userDetailsService.loadUserByUsername(getUsername(token))
        return UsernamePasswordAuthenticationToken(userDetails, "", userDetails.authorities)
    }

    fun getUsername(token: String?): String {
        return Jwts.parser().setSigningKey(ENCODED_SECRET_KEY).parseClaimsJws(token).body.subject
    }

    fun extractToken(req: HttpServletRequest): String? {
        return req.getHeader(TOKEN_HEADER_NAME)
    }

    fun extractToken(accessor: StompHeaderAccessor): String? {
        return accessor.getNativeHeader(TOKEN_HEADER_NAME)?.get(0)
    }

    fun validateToken(token: String?): Boolean {
        return try {
            val claims = Jwts.parser().setSigningKey(ENCODED_SECRET_KEY).parseClaimsJws(token)
            !claims.body.expiration.before(Date())
        } catch (e: JwtException) {
            throw JwtException("Expired or invalid JWT token")
        } catch (e: IllegalArgumentException) {
            throw JwtException("Expired or invalid JWT token")
        }
    }

    companion object {
        private const val ROLES = "BASIC_ROLE"
        private const val SECRET_KEY = "NON_SAFE_SECRET"
        private val ENCODED_SECRET_KEY = Base64.getEncoder().encodeToString(SECRET_KEY.toByteArray())
        private const val TOKEN_HEADER_NAME = "token"
        private const val VALID_TIME_MS = 3600000
    }
}